### ISSUE TITLE
aggregations and Ferret's limit on the number of datasets

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1649]
**Reported by @AndrewWittenberg on 7 Mar 2016 20:33 UTC**
Working with the new T-E-F aggregations, I'm very quickly running up against Ferret's limit on the max number of open datasets (maxdsets=100).

For example, an F-E aggregation of just 4 years of monthly forecasts, just 2 members each, uses up (48+1)+(48+1)+1 = 99 datasets.  (Here the "+1"'s represent the the intermediate F aggregation dataset, and the final ensemble aggregation dataset.

Aggregating GFDL's full set of ensemble forecasts, say 12 members for 1960-2016, for just a single physical variable (say t_surf), would use up ~10k datasets.

How shall we approach this issue?

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/2377

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 7 Mar 2016 22:55 UTC**
Glad that you are taking the new capabilities for a ride.

I'm aware of the limitations that the current maxdsets imposes.  I can up this maxdset limit to (say) around 500 by just changing 2 parameters in the code and recompiling.  A bit of code smithing is desirable to go past that point;  I need to eliminate the four variables that are double dimensioned in maxdset and maxvars.  Order(4 hours) of work to do so (and the code would be made more efficient in the process.)


There's nothing fundamental in the architecture of Ferret that would prevent us from having 10K datasets.  The memory overhead would remain a tiny fraction of the memory needed to wrestle the floating point arrays in and out.  I suspect we'll also quickly be wanting a new level of subtlety in SHOW DATA (& co.) so one doesn't get lost in the tangle of aggregation members.

In short, I'm not worried, but I agree somethign needs to be done.  I'll up the dataset limit to a few hundred right away.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 8 Mar 2016 21:22 UTC**
A related issue is all the axes that are defined for the constituent datasets, e.g. in an FMRC aggregation.  We might run into limits on the number of axes, as well as needing a new level of subtlety in the SHOW AXIS command.

Perhaps also limits on the number of grids?

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 8 Mar 2016 22:37 UTC**
These problems haven't shown up yet, but possible.

WRT axes, the only proliferation is 1 time axis per file in an FMRC.  ENSEMBLE aggregations (and X,Y,Z of FMRCs) should re-use the shared member axes.  When a TSERIES is defined the sub-axes of the individual files are not stored -- just the one, full time axis of the aggregation.  If it was *really* important, the time axes of FMRC members could probably be freed up at the time that the /HIDE of the FMRC command was processed, though doing so would open several middle sized cans of worms.

Grids are unlikely to be a problem.  A grid definition is hardly more than the 6 pointers to its axes, so one could have 100K grids bouncing around in Ferret and never really know the difference.

In summary: of the scalability challenges to working with T-E-F aggregations these aren't the ones I am worried about.  More concern to me is 1) number of datasets (which we know how to address); and 2) geometrically expanded sizes of calculations.  For example suppose if, as in the demo I showed when we visited GFDL last year, one wants to make a global map showing the ensemble average of the calendar-time averaged forecast lead time at which the forecast reaches a certain threshold accuracy.  The total data that goes into this calculation is nX*nY*nT*nE*nF.  It can be frickin huge!   The T, E and F axes all get compressed to points during this analysis.  Ferret's internal logic for breaking up calculations protects axes that are going to be compressed (because some compressing calculations need to see all of the data on the axis at once).  Which means this calculation gets broken up along Y.  The result is pretty inefficient IO -- each file gets opened and accessed multiple times during the analysis. (On the other hand, Ferret will at least succeed with the calculation ... and its not an easy calculation to automate.)

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 8 Mar 2016 22:54 UTC**
Yes, the datasets will be huge!  But I think that's just the nature of the problem.  At least now we have a way to describe and poke around these datasets with Ferret -- to guide us in setting up a possibly huge calculation to be done overnight, possibly with multiple Ferret sessions running on separate processors.

**Comment by AnsleyManke** [comment]
This was addressed with the fix to issue #1694.  The number of datasets is now 5000.  And the number of grids is 10000, see #1657.
